Cross-DAO Redemptions

Procrat opened the pull request:
This PR associates redemptions with the user, rather than with a DAO. The related design document can be found [here](https://docs.google.com/document/d/1-x2-pDn2jTHGbkGdiPcPi9s0_GXeehOq7yEAs8n0mL8/edit).

- [x] All redemptions page
  - [x] Button to redeem all rewards
- [x] Redemptions button in the header
- [x] Remove per DAO redemption button and page
- [x] Redemptions quick menu

Procrat commented:
I'll rebase this PR later today, but other than that, it's up for review.

Procrat commented:
I've rebased it now.

Be sure to have a close look at 404fe03. It's unrelated to the goal of this PR, but I noticed so much unused CSS classes that I thought it would help future development to clean that up a bit.

jellegerbrandy commented:
@Procrat many thanks for the PR :muscle: :fire: 

i've gone over the code very quickly, but it probably should be first reviewed by our product manager, @A-Zak .  After he gives the ok for the functionality, I'll review in more detail

There is a preview app on heroku here for this PR: https://alchemy-staging-rinkeb-pr-1047.herokuapp.com/ (talking to rinkeby)

Also thanks for the cleanups and the CSS, much appreciated :-). Perhaps next time, if there will be one,  this can be a separate PR?

Procrat commented:
@jellegerbrandy Thanks for that. And yeah, I'll be sure to make a separate PR for separate things next time. Sorry about that.

It sounded like @A-Zak would be busy with Berlin stuff this week, but I'm in no rush. :-)

Just to make sure we're on the same page, this is the implementation of [this Genesis proposal](https://alchemy.daostack.io/dao/0x294f999356ed03347c7a23bcbcf8d33fa41dc830/proposal/0x39b3c5cf97197765f2afa7068f31f295bfc6ddb8232281f52db055b3093a5ea0), and the specs and designs I linked above are what Alex sent us.

tibetsprague commented on src/components/Redemptions/RedemptionsPage.tsx:
```diff
@@ -0,0 +1,291 @@
+  redeemProposal: arcActions.redeemProposal,
+};
+
+class FallibleRedemptionsPage extends React.Component<IStateProps & IDispatchProps, null> {
```
Fallible?

Procrat replied:
I thought it would make sense to make a wrapping component (FallibleRedemptionsPage) to deal with the data fetching, which can possibly fail, and an inner component (RedemptionsPage), which shouldn't.

Is this a question/remark about the name or about the abstraction? Happy to change either if you have an idea.

tibetsprague replied:
Oh just the name. What about Redemption Data? Or RedemptionaPageData?

Procrat replied:
I've changed the name in the last rebase to `SubscribedRedemptionsPage` to make it consistent with other components. That's better, isn't it?

dkent600 commented:
If you haven't any rewards, then the popup should feedback with a text message, not just an empty window.

![image](https://user-images.githubusercontent.com/1821666/63595535-31e72700-c587-11e9-963f-3e86e1fae8c3.png)

dkent600 commented:
This window should not have a fixed height as it does, rather should fit the required vertical space, with some maximum above which there appears a scroll bar.

dkent600 commented:
Could be nice if the star icon and red round circle thingy somehow could display the number of available redemptions.

dkent600 commented:
For the Ganache 0 account I'm seeing "Loading..." that doesn't go away:

![image](https://user-images.githubusercontent.com/1821666/63596247-cdc56280-c588-11e9-961a-f61faf49b07d.png)

dkent600 commented:
Why do some items have a redeem button and others don't?

![image](https://user-images.githubusercontent.com/1821666/63596415-36144400-c589-11e9-8f81-a48583feb657.png)

dkent600 commented:
I don't think we need a star button that hover-expands to a Redeem button.  There is plenty of space here, and it is easier for the user, to just have a "Redeem" button.

dkent600 commented:
Why do we have a "View All" link?  Does this window not already contain all redemptions?

![image](https://user-images.githubusercontent.com/1821666/63596573-8b505580-c589-11e9-8049-6661908131e8.png)

dkent600 commented:
When I click on "View All" with the Ganache 0 account, I just get an error message:

> Report feedback
> Error: No action matching these callData could be found

dkent600 commented:
Bit of a layout problem here on mobile ("responsive" format):

![image](https://user-images.githubusercontent.com/1821666/63597462-82f91a00-c58b-11e9-98bf-398c2d865f31.png)

Procrat commented:
> this is going to be fun to refactor with the new better performing withSubscriptions Higher Order Component now merged into dev to replace :)

It took a bit to rebase, but that new HOC does make things quite a bit nicer!

Procrat commented:
> The star menu icon should have a hover Tooltip explaining what it does ("Redeem your rewards" or something like that).

> If you haven't any rewards, then the popup should feedback with a text message, not just an empty window. And maybe there is no need for a "View All" link.

> Could be nice if the star icon and red round circle thingy somehow could display the number of available redemptions.

> Why do we have a "View All" link? Does this window not already contain all redemptions? Does this link need a better description?

> This window should not have a fixed height, or large minHeight, as it does, rather should fit the required vertical space, with some maximum above which there appears a scroll bar.

For these remarks, I suggest discussing with @A-Zak, since they weren't really specified in the original design.

> For the Ganache 0 account I'm seeing "Loading..." that doesn't go away:
Which account are you referring to?

> Why do some items have a redeem button and others don't?
In the interest of code reuse, I reused the ActionButton component that was already used in the proposal cards. I was under the assumption that if the GraphQL query would return it as a redeemable proposal, the redeem button would always show up. Can you think of a reason why that isn't the case there?

> I don't think we need a star button that hover-expands to a Redeem button. There is plenty of space here, and it is easier for the user, to just have a "Redeem" button.
Right, I'll look into that.

> Bit of a layout problem here on mobile ("responsive" format). Could be this would be fixed by #1081
I've just noticed that as well after a rebase. Could it be that this "connect" button while being logged in is something new? I thought being logged in and being connected to your wallet meant the same thing, but perhaps I'm wrong in assuming that? What would the best solution be here UI-wise?

dkent600 commented:
>> Bit of a layout problem here on mobile ("responsive" format). Could be this would be fixed by #1081

> I've just noticed that as well after a rebase. Could it be that this "connect" button while being logged in is something new? I thought being logged in and being connected to your wallet meant the same thing, but perhaps I was wrong in assuming that? What would the best solution be here UI-wise?

@Procrat How about you get the latest from the dev branch where the mobile layout of the Connect button is fixed, then see what it looks like?  Go from there...

The term "Connect" is used when we have a cached current account, obtained from localStorage, but not a web3 provider for it that enables us to send txs on its behalf.

dkent600 commented:
>> For the Ganache 0 account I'm seeing "Loading..." that doesn't go away:

> Which account are you referring to?

@Procrat `accounts[0]`

dkent600 commented:
>> Why do some items have a redeem button and others don't?

> In the interest of code reuse, I reused the ActionButton component that was already used in the proposal cards. I was under the assumption that if the GraphQL query would return it as a redeemable proposal, the redeem button would always show up. Can you think of a reason why that isn't the case there?

@Procrat I will just refer you to the `ActionButton` code to figure out why it is behaving that way.  I think this needs to be understood since it makes no sense for a proposal to appear in your list if it can't be redeemed.

dkent600 commented:
> For these remarks, I suggest discussing with @A-Zak, since they weren't really specified in the original design.

@A-Zak  Can you please review these comments, thanks!

Procrat commented:
In the meantime, I've had a chat with Alex and received some UI mockups about the unspecified bits. I've rebased and
  - [x] put the redemptions button on the left for mobile displays,
  - [x] always let the text show on the redeem buttons in the menu, and
  - [x] made the Alchemy logo clickable (so that the user can navigate away from the redemptions page).

> get the latest from the dev branch where the mobile layout of the Connect button is fixed, then see what it looks like? Go from there...

Since the button is on the left now, this is fortunately no longer an issue.

> @Procrat ganache `accounts[0]` in the preview app against rinkeby.

I'm afraid the preview app seems to be down. I'm still unclear on which account you're referring to, but perhaps seeing the preview app would have made it clear?

> > In the interest of code reuse, I reused the ActionButton component that was already used in the proposal cards. I was under the assumption that if the GraphQL query would return it as a redeemable proposal, the redeem button would always show up. Can you think of a reason why that isn't the case there?
> 
> @Procrat I will just refer you to the `ActionButton` code to figure out why it is behaving that way. I think this needs to be understood since it makes no sense for a proposal to appear in your list if it can't be redeemed.

I was just curious whether you had an inkling about something that could have been easily overlooked. Anyway, looking at the ActionButton code, it shows the redeem button when `proposal.executedAt && proposal.accountsWithUnclaimedRewards.includes(currentAccountAddress.toLowerCase())` whereas the items in the menu are fetched with the GQL query `{ proposals(where: { accountsWithUnclaimedRewards_contains: ["${currentAccountAddress}"] }) { id } }`. Assuming that account addresses are lower case, that would mean that an account can occur in the `accountsWithUnclaimedRewards` list while the proposal has not (yet) been executed.

Venturing into sugraph code, it looks like only `insertGPRewards` in `src/domain/reward.ts` can add to this `accountsWithUnclaimedRewards` list and that function is only called when the proposal is executed or expired in the queue. Could that be what happened in your case then?

Just to be clear, that GQL was taken from the existing per-DAO redemption page, so I would assume that whatever bug shows up here, is already present in the current version of Alchemy. Can you confirm that?

Procrat commented:
I've
  - [x] ensured the menu is only maximum as heigh as the viewport (with some margin), and
  - [x] changed the menu to look like [this mockup](https://app.zeplin.io/project/5c4de2851386a637a927c270/screen/5ca1f402bb1e6105d51575d0) when there's nothing to redeem.

Apart from that issue that @dkent600 encountered (which I'm unclear on how to reproduce), are there any other issues left before this can be merged?

Procrat commented:
I've added a tiny bit of robustness to `RedemptionsString` (which was apparently a component that wasn't being used at the moment).

I've encountered a bug with account 0x90f8bf6a... (one of the test accounts mentioned in the README) on the preview app, where this error was thrown: "Error: No action matching these callData could be found". After some digging, it looks like this is an issue with a particular DutchX proposal and this was apparently already an issue before this PR: see for example [this other PR preview](https://alchemy-staging-rinkeb-pr-1097.herokuapp.com/dao/0x1eb3bcae24d42fde2f90e91d26313ac3946c4764/redemptions/).

jellegerbrandy commented:
hi @Procrat , could you make an issue for the preview app error and say a bit more about the circumstances?

A-Zak commented:
@Procrat So 1st of all thank you so much for the great work (and extra effort) - it looks really good :)

1st things 1st - ganache 0 account is the account with this private key `4F3EDF983AC636A65A842CE7C78D9AA706D3B113BCE9C46F30D7D21715B23B1D`.
I'm not sure exactly where they come from, but I think these accounts come from the default ganache keychain (which is the same for everyone), so we often use them for debugging... ^_^

- [ ] And I can confirm that if you connect the ganache 0 account to the preview app (which is back online 💪) and try to open the redemptions dropdown - alchemy crashes... Same if you go to the redemptions page directly :)
So it would be great if you could take a look at that... :)

-----

- [ ] When having a lot of redemptions, the dropdown only loads the **last** few redemptions, the rest stay like in "loading" /:
![redemptions dropdown](https://user-images.githubusercontent.com/302774/64556224-aec81e00-d347-11e9-9803-d5e7a5ced3f7.gif)

This happened on my personal account... To test this you can edit the `current account` in the local storage to "0x6d1A4be65723355a9f5b12C446780a5000849461" :)
![image](https://user-images.githubusercontent.com/302774/64556601-6bba7a80-d348-11e9-8dfd-4ec7d94a1b00.png)

-----

- [ ] Pressing the `redeem all` button when not connected is not doing anything. if not connected, it should prompt the user to connect to a wallet... Most of the other buttons have this behaviour... ^_^

Procrat commented:
@A-Zak Thanks for the review!

I didn't realise those accounts were common across projects using Ganache. Thanks for pointing that out. Those were the ones I've been testing with anyway.

With regards to Alchemy crashing, that was the bug [I had pointed out earlier](https://github.com/daostack/alchemy/pull/1047#issuecomment-528307818). I've created [an issue](https://github.com/daostack/alchemy/issues/1122) with my findings.

I've rebased and fixed the other two issues you mentioned as well.

Let me know if anything else needs to happen!

A-Zak commented:
Hey, @Procrat this looks really great thank you! :)
I went over all the fixes, and also did an overall check and everything looks great except one tiny thing - the redeem all button on the all redemptions page is still not reacting if the user is not connected...

Other than this everything is good to go! 💪

A-Zak commented:
Also, it looks like there are a few conflicts to resolve... it'd be great if you could clear those as well 🙏

Procrat commented:
@A-Zak: Oh, I had only done it for the "redeem all" button in the menu and had forgotten about the other one. Should be fixed now!

Procrat commented:
(Just another rebase)

A-Zak commented:
This looks great! :)
@jellegerbrandy @tibetsprague @dkent600 - did you look at the code itself? if so let's merge it in ^_^

tibetsprague commented on src/components/Redemptions/RedemptionsPage.tsx:
```diff
@@ -0,0 +1,269 @@
+        return combineLatest(Array.from(daoIds, (id: any) => arc.dao(id).state()))
+          .pipe(defaultIfEmpty([]));
+      }));
+    return combineLatest(daos, proposals);
```
is there any reason we actually need to to subscribe to the daos? i bet @jellegerbrandy has something to say about how we can reduce the number of subscriptions here given all the work he has been doing on that.

Procrat replied:
The proposal cards on the page require a DAO state as a prop. It looks like various subcomponents use it: ProposalData, ActionButton, VoteButtons, StakeButtons, AccountPopup, etc.
